AdminServer/appscale/admin/instance_manager/instance_manager.py

```diff
@@ -0,0 +1,610 @@
+
+    yield self._monit_operator.reload(self._thread_pool)
+    yield self._clean_old_sources()
+
+  def _get_lowest_port(self):
+    """ Determines the lowest usuable port for a new instance.
+
+    Returns:
+      An integer specifying a free port.
+    """
+    existing_ports = {instance.port for instance in self._running_instances}
+    port = STARTING_INSTANCE_PORT
+    while True:
+      if port in existing_ports:
+        port += 1
+        continue
+
+      return port
+
+  @gen.coroutine
+  def _fulfill_assignments(self):
+    """ Starts and stops instances in order to fulfill assignments. """
+
+    # Wait until either the grooming interval has passed or the assignments
+    # have been updated.
+    try:
+      yield gen.with_timeout(datetime.timedelta(seconds=self.GROOMING_INTERVAL),
+                             self._update_event.wait())
+    except TimeoutError:
+      pass
+
+    self._update_event.clear()
+
+    if time.time() > self._last_cleanup_time + INSTANCE_CLEANUP_INTERVAL:
+      yield self.stop_failed_instances()
+
+    # Stop versions that aren't assigned.
+    to_stop = [instance for instance in self._running_instances
+               if instance.version_key not in self._assignments]
+    for instance in to_stop:
+      yield self._stop_app_instance(instance)
+
+    for version_key, assigned_ports in self._assignments.items():
+      project_id, service_id, version_id = version_key.split(
+        VERSION_PATH_SEPARATOR)
+      version_instances = [instance for instance in self._running_instances
+                           if instance.version_key == version_key]
+      try:
+        version = self._projects_manager[project_id][service_id][version_id]
+      except KeyError:
+        # If the version node no longer exists, stop any running instances.
+        for instance in version_instances:
+          yield self._stop_app_instance(instance)
+
+        continue
+
+      # Ensure version updates trigger fulfillment work.
+      version.add_callback(self._handle_version_update)
+
+      # Stop instances that aren't assigned.
+      for instance in version_instances:
+        if instance.port not in assigned_ports:
+          yield self._stop_app_instance(instance)
+
+      # Start assigned instances that aren't running.
+      for port in assigned_ports:
+        if port == -1:
+          port = self._get_lowest_port()
+
+        instance = Instance(version.revision_key, port)
+        if instance not in version_instances:
+          self._start_instance(version, instance.port)
+
+      # If there are any instances running an older revision, restart them.
+      for instance in version_instances:
+        if instance.revision_key != version.revision_key:
+          yield self._stop_app_instance(instance)
+          yield self._start_instance(version, instance.port)
+
+  @gen.coroutine
+  def _fulfillment_loop(self):
+    """ Continually tries to fulfill assignments. """
+    while True:
```

Comment:
I see benefits of using AsyncEvent with endless loop blocked by the event.
But it took ~20 minutes for me to understand how it works=)
Asynchronous code is complicated by its nature. I don't have strong feeling that my approach is better, but anyway I suggest to consider it.

So currently, from what I understood, the flow is (I'll mention only relevant steps):
 1. Initialization: go through assignments declared in ZK, add loop-unlocking callback to every version.
 2. Do idle iterations every GROOMING_INTERVAL
 3. Version is updated in ZK
 4. Version manager is notified about it and does update
 5. Version manager calls loop-unlocking callback
 6. _fulfill_assignments locks loop again
 7. _fulfill_assignments ensures that all versions will unlock endless loop when new revision arrives.

Would _fulfill_assignments be "notified" when **new** version is deployed? (read: Is AC state node updated when new version is deployed)?

My concern is that it looks like InstanceManager is simply subscribed to version updates. This subscription is done by endless loop + asyncevent + callback, which are all comparably complicated tools.

Wouldn't it be better to implement subscription mechanism in GlobalProject manager to raise updates from version level to global project and then to anythink interested in any updates inside projects?
In this case InstanceManager would subscribe once plus add periodicalcallback with GROOMING_INTERVAL.

Reply:
@jeanleonov First of all, thanks for the speedy review of these extensive changes! I agree that this could be a lot clearer. I haven't used async events much, so I'm still trying to learn the best way to use them.

There's a slight difference difference between your description and what I've intended. The machine's assignments come from `/appcontroller/state`, but the version details come from `...services/<service-id>/versions/<version-id>`.

The InstanceManager does not care about changes to versions that aren't assigned to it. So when a new version is deployed, the fulfillment work should not be triggered yet. If the scheduler later assigns some of that new version's instances to the relevant machine, it would update `/appcontroller/state` and trigger the fulfillment work. Similarly, the InstanceManager does not need to perform fulfillment work if a version that it is not assigned to run gets a new revision or deleted, etc.

I am 100% in favor of code readability over efficiency here, but I don't want to mislead the reader into thinking that the InstanceManager cares about _any_ changes to the configured projects. I'll try to make some changes to clarify that fulfillment work is triggered by any of 3 conditions (grooming interval, update to the machine's assignments, and an update to an assigned version).

To address your last concern, the subscription to version changes is idempotent. I wanted to avoid a situation where a version object is deleted and later recreated, losing the callback. I'm going to move the subscription work away from the fulfillment loop, though. I agree that it's a bit weird right now.

Your idea about subscribing directly to the projects manager is a good one. I'll give it a shot.